python/cudf/cudf/core/frame.py

```diff
@@ -1324,10 +1165,18 @@ def fillna(
             else:
                 filled_data[col_name] = col.copy(deep=True)
 
-        return self._mimic_inplace(
-            self._from_data(data=filled_data, index=self._index),
-            inplace=inplace,
+        # TODO: This logic needs to move into the IndexedFrame class.
+        old_index = self._index
+        ret = self._mimic_inplace(
+            self._from_data(data=filled_data), inplace=inplace,
         )
+        # TODO: Split this logic into the IndexedFrame class.
+        if isinstance(self, cudf.core.indexed_frame.IndexedFrame):
+            if inplace:
+                self._index = old_index
+            else:
+                ret._index = old_index
```

Comment:
Actually we can't do this because we are calling `_mimic_inplace` before this and in the inplace case we still need to replace the old_index back into self.

Reply:
That means we are tampering with `self._index` in `_mimic_inplace` if `inplace` is `True`? That seems weird no?

Reply:
Confirmed offline that this is fine.